src/libopenrave/kinbodygrab.cpp

```diff
@@ -18,32 +18,51 @@
 
 namespace OpenRAVE {
 
-/// \brief Push link to listNonCollidingLinksWhenGrabbed, only if it has no collision with the whole grabbedBody.
-static void _PushLinkIfNonColliding(std::list<KinBody::LinkConstPtr>& listNonCollidingLinksWhenGrabbed,
+/// \brief Push link to listNonCollidingLinkPairs if the given links are not in collision each other. This is for grabbed-grabber link pairs.
+static void _PushLinkIfNonColliding(std::list<std::pair<KinBody::LinkConstPtr, KinBody::LinkConstPtr> >& listNonCollidingLinkPairs,
                                     CollisionCheckerBasePtr& pchecker,
-                                    const KinBody::LinkPtr& pLinkToCheck, const KinBody& grabbedBody)
+                                    const KinBody::LinkPtr& pGrabberLinkToCheck, const KinBody& grabbedBody)
 {
-    KinBody::LinkConstPtr pLinkToCheckConst(pLinkToCheck);
+    KinBody::LinkConstPtr pGrabberLinkToCheckConst(pGrabberLinkToCheck);
     for (const KinBody::LinkPtr& pGrabbedBodylink : grabbedBody.GetLinks()) {
-        if( pchecker->CheckCollision(pLinkToCheckConst, KinBody::LinkConstPtr(pGrabbedBodylink)) ) {
-            return; // if colliding, do not push.
+        if( !pchecker->CheckCollision(pGrabberLinkToCheckConst, KinBody::LinkConstPtr(pGrabbedBodylink)) ) {
+            listNonCollidingLinkPairs.emplace_back(pGrabbedBodylink, pGrabberLinkToCheck);
         }
     }
-    // if not colliding with any of links in grabbedBody, push it.
-    listNonCollidingLinksWhenGrabbed.push_back(pLinkToCheck);
 }
 
-/// \brief remove link from listNonCollidingLinksWhenGrabbed if its parent is same as the given body.
-template<typename KinBodyPtrT>
-static void _RemoveLinkFromListNonCollidingLinksWhenGrabbed(std::list<KinBody::LinkConstPtr>& listNonCollidingLinksWhenGrabbed,
-                                                            const KinBodyPtrT& pGrabbedBody)
+/// \brief Check if link pair is included in the list. Note that envBodyIndex for pLink1ToSearch's parent KinBody should be smaller than envBodyIndex for pLink2ToSearch's parent KinBody.
+/// \param[in] pLink1ToSearch, pLink2ToSearch : ptr of links to check.
+/// \param[in] listNonCollidingLinkPairs : taret list.
+static bool _IsLinkPairIncluded(const KinBody::Link* pLink1ToSearch,
+                                const KinBody::Link* pLink2ToSearch,
+                                const std::list<std::pair<KinBody::LinkConstPtr, KinBody::LinkConstPtr> >& listNonCollidingLinkPairs)
```

Comment:
```suggestion
/// \brief Return true if the link pair (pLink1ToSearch, pLink2ToSearch) is already in the given
///        listNonCollidingLinkPairs. Note that the order of the links must be such that body1.GetEnvironmentBodyIndex()
///        < body2.GetEnvironmentBodyIndex() where body1 is the kinbody pLink1ToSearch belongs to and body2 is the
///        kinbody pLink2ToSearch belongs to.
/// \param[in] pLink1ToSearch, pLink2ToSearch : ptr of links to check.
/// \param[in] listNonCollidingLinkPairs : taret list.
static bool _IsLinkPairIncluded(const KinBody::Link* pLink1ToSearch,
                                const KinBody::Link* pLink2ToSearch,
                                const KinBody::ListNonCollidingLinkPairs& listNonCollidingLinkPairs)
```

Reply:
thanks!

since there was conflict with other commit, i manually applied this improvement : https://github.com/rdiankov/openrave/pull/1438/commits/455ed5f9b2d381cefefa5e79543e592feae0b8b9